File: src/main/java/at/hrechny/predictionsbot/mapper/SeasonMapper.java
package at.hrechny.predictionsbot.mapper;

import at.hrechny.predictionsbot.database.entity.CompetitionEntity;
import at.hrechny.predictionsbot.database.entity.SeasonEntity;
import at.hrechny.predictionsbot.model.Season;
import org.mapstruct.Mapper;
import org.mapstruct.Mapping;
import org.mapstruct.MappingTarget;

@Mapper(uses = YearMapper.class)
public interface SeasonMapper {

  @Mapping(target = "competition", source = "competition.name")
  Season entityToModel(SeasonEntity source);

  @Mapping(target = "rounds", ignore = true)
  @Mapping(target = "competition", ignore = true)
  void updateEntity(@MappingTarget SeasonEntity entity, Season model);

  @Mapping(target = "id", source = "season.id")
  @Mapping(target = "competition", source = "competition")
  @Mapping(target = "rounds", ignore = true)
  SeasonEntity modelToEntity(CompetitionEntity competition, Season season);

}

File: src/main/kotlin/at/hrechny/predictionsbot/mapper/SeasonMapper.kt
package at.hrechny.predictionsbot.mapper

import at.hrechny.predictionsbot.database.entity.CompetitionEntity
import at.hrechny.predictionsbot.database.entity.SeasonEntity
import at.hrechny.predictionsbot.controller.model.competition.SeasonCreateRequestDto
import at.hrechny.predictionsbot.controller.model.competition.SeasonResponseDto
import at.hrechny.predictionsbot.controller.model.competition.SeasonUpdateRequestDto
import jakarta.inject.Singleton

@Singleton
class SeasonMapper(
    private val yearMapper: YearMapper,
) {
    fun entityToModel(source: SeasonEntity): SeasonResponseDto =
        SeasonResponseDto().apply {
            id = source.id
            year = yearMapper.asYear(source.year)
            competition = source.competition?.name
            active = source.active
        }

    fun updateEntity(entity: SeasonEntity, model: SeasonUpdateRequestDto) {
        if (model.id != null) {
            entity.id = model.id
        }
        if (model.year != null) {
            entity.year = yearMapper.asString(model.year)
        }
        entity.active = model.active
    }

    fun modelToEntity(competition: CompetitionEntity, season: SeasonCreateRequestDto): SeasonEntity =
        SeasonEntity().apply {
            id = season.id
            this.competition = competition
            year = yearMapper.asString(season.year)
            active = season.active
        }
}
